### PR TITLE
Update addAnnotation methods to use SemanticallyEqual

### PR DESCRIPTION
**Pull request by SofiaBrittoSchwartz**
Created public versions of the addAnnotation methods and created a public static buildAnnotation method in J.Annotation for ease of access.

- Updated AddAnnotation to use SemanticallyEqual to prevent duplicate annotations

### CODE REVIEW

**Comment by jkschneider** [comment]
@SofiaBrittoSchwartz For better discovery, could we place these as non-static methods on `J.MethodDecl` and `J.ClassDecl` respectively?

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java [hunk 1]
```diff
@@ -1053,6 +1056,46 @@ public EnumValueSet getEnumValues() {
             return new FindType(clazz).visit(this);
         }
 
+        public static J.ClassDecl addAnnotation(
+                J.ClassDecl c,
```
Comment (jkschneider) [inline rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:1060]: You can remove this parameter `c`, as it should be able to refer to `this`. Suggestion is to make this non-static.

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java [hunk 2]
```diff
@@ -1053,6 +1056,46 @@ public EnumValueSet getEnumValues() {
...
+                boolean isTopLevelClass,
```
Comment (jkschneider) [inline rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:1061]: You should be able to simplify and remove this now. Add the annotation and then apply `AutoFormat`. It's only used to detect a special case of formatting that should be caught in `AutoFormat`.

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java [hunk 3]
```diff
@@ -1053,6 +1056,46 @@ public EnumValueSet getEnumValues() {
...
+                JavaType.Class annotationType,
+                List<Expression> arguments,
+                JavaFormatter formatter
```
Comment (jkschneider) [inline rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:1064]: `AutoFormat` does not need to have a `JavaFormatter` passed to it.

**Review by jkschneider — CHANGES_REQUESTED** [review]

**Review by jkschneider — APPROVED** [review]

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java [hunk 4]
```diff
@@ -2046,6 +2070,32 @@ public boolean hasType(String clazz) {
             return new HasType(clazz).visit(this);
         }
 
+        public J.MethodDecl addAnnotation(
+                J.MethodDecl m,
+                JavaType.Class annotationType,
+                List<Expression> arguments,
+                JavaFormatter formatter
+        ) {
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2078]: Same question here as I had about `J.ClassDecl.addAnnotation()`

**Review by sambsnyd — CHANGES_REQUESTED** [review]
Good work!

File: rewrite-java/src/main/java/org/openrewrite/java/AddAnnotation.java [hunk 5]
```diff
@@ -142,38 +119,33 @@ public Scoped(Tree scope, String annotationTypeName, Expression... arguments) {
...
-                            m = m.withName(m.getName().withPrefix(prefix));
-                        }
-                    }
+                Formatting formatting = !m.getAnnotations().isEmpty() ?
+                        m.getAnnotations().get(0).getFormatting() :
+                        !m.getModifiers().isEmpty() ?
+                                m.getModifiers().get(0).getFormatting() :
+                                m.getReturnTypeExpr() != null ?
+                                    m.getReturnTypeExpr().getFormatting() :
+                                    m.getName().getFormatting();
```
[the remark this answers is not in the document]
Comment (SofiaBrittoSchwartz) [reply rewrite-java/src/main/java/org/openrewrite/java/AddAnnotation.java]: Refactored to an if/else statement

**Review by sambsnyd — APPROVED** [review]

**Pull request merged**